src/libraries/Native/Unix/System.Native/pal_random.c

```diff
@@ -20,91 +21,230 @@ Generate random bytes. The generated bytes are not cryptographically strong.
 
 */
 
-void SystemNative_GetNonCryptographicallySecureRandomBytes(uint8_t* buffer, int32_t bufferLength)
+#if HAVE_ARC4RANDOM_BUF
+
+#define MAY_NEED_XOR_WITH_LRAND 0
+static bool PlatformRandomFillBuffer(uint8_t *buffer, size_t bufferLength)
 {
-    assert(buffer != NULL);
+    arc4random_buf(buffer, bufferLength);
+    return true;
+}
 
-#if HAVE_ARC4RANDOM_BUF
-    arc4random_buf(buffer, (size_t)bufferLength);
-#else
-    static volatile int rand_des = -1;
-    long num = 0;
-    static bool sMissingDevURandom;
-    static bool sInitializedMRand;
+#elif HAVE_SYS_GETRANDOM
 
-    if (!sMissingDevURandom)
+#define MAY_NEED_XOR_WITH_LRAND 1
+static bool PlatformRandomFillBuffer(uint8_t *buffer, size_t bufferLength)
+{
+    static bool sSyscallNotSupported;
+
+    if (sSyscallNotSupported)
     {
-        if (rand_des == -1)
-        {
-            int fd;
+        return false;
+    }
+
+    for (;;)
+    {
+        long bytesFilled = getrandom(buffer, bufferLength, 1 /* GRND_NONBLOCK */);
 
-            do
+        if (bytesFilled == -1)
+        {
+            switch (errno)
             {
+                case EINTR:
+                    continue;
+                case ENOSYS:
+                    sSyscallNotSupported = true;
+                    break;
```

Comment:
Break? Or return false?  My brain-flow analysis says that right now the first caller to get ENOSYS goes into an infinite loop.

Reply:
Good catch. Will fix in the next push.